# SQL Project

The coronavirus took the world by surprise, changing the usual order of things. In their free time, residents of cities no longer go outside, do not visit cafes and shopping centers. But there is more time for books. Startups noticed this and rushed to create apps for those who like to read.

Our company decided to be on the wave and bought a large subscription-based book reading service.

Tasks:  

 - to **analyze the database**. It contains information about books, publishers, authors, as well as user reviews of books and
 - **formulate a proposition** for a new product.

# Data description
The **books** table

Contains the data about the books:

- `book_id` — book ID;
- `author_id` — the author's ID;
- `title` — the title of the book;
- `num_pages` — number of pages;
- `publication_date` — date of publication of the book


The **authors** table

Contains the information about the authors:

- `author_id` — the author's ID;
- `author` — the author's name.


The **publishers** table

Contains the data about publishing houses:

- `publisher_id` — id of the publisher;
- `publisher` — the name of the publisher;


The **ratings** table

Contains the data about user ratings of books:

- `rating_id` — rating ID;
- `book_id` — the id of the book;
- `username` — the name of the user who left the rating;
- `rating` — rating of the book.


The **reviews** table

Contains the data about user reviews of the books:

- `review_id` — review ID;
- `book_id` — book ID;
- `username` — the name of the user who wrote the review;
- `text` — the text of the review.


## Connecting to the database

In [12]:
# importing the library
import pandas as pd
from sqlalchemy import create_engine


# setting the parameters
db_config = {'user': 'praktikum_student', # username
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # connecting port
             'db': 'data-analyst-final-project-db'} # the name of the database
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])


# saving the connector
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## Table analysis

### books

In [13]:
# getting the books table
query = ''' 

SELECT *
FROM books
LIMIT 5 

'''
books = pd.io.sql.read_sql(query, con = engine) 
display(books)
books.dtypes

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

book_id              int64
author_id            int64
title               object
num_pages            int64
publication_date    object
publisher_id         int64
dtype: object

In the books table we see 6 columns: `book_id` — book id; `author_id` — author id; `title` — book title; `num_pages` — number of pages; `publication_date` — publication date of the book; `publisher_id` — publisher id.

### authors

In [14]:
# getting the authors table
query = ''' 

SELECT *
FROM authors
LIMIT 5 

'''
authors = pd.io.sql.read_sql(query, con = engine) 
display(authors)
authors.dtypes

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

author_id     int64
author       object
dtype: object

The authors table contains the columns `publisher_id` — the ID of the publisher; `publisher` — the name of the publisher.

### publishers

In [15]:
# getting the publishers table
query = ''' 

SELECT *
FROM publishers
LIMIT 5 

'''
publishers = pd.io.sql.read_sql(query, con = engine) 
display(publishers)
publishers.dtypes

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

publisher_id     int64
publisher       object
dtype: object

There are 2 columns in the table: `publisher_id` is the ID of the publisher; `publisher` is the name of the publisher.

### ratings

In [16]:
# getting the ratings table
query = ''' 

SELECT *
FROM ratings
LIMIT 5 

'''
ratings = pd.io.sql.read_sql(query, con = engine) 
display(ratings)
ratings.dtypes

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

rating_id     int64
book_id       int64
username     object
rating        int64
dtype: object

In the table there are the following columns:  `rating_id` - the rating ID; `book_id` — the book id; `username` — the name of the user who left the rating; `rating` — the rating of the book.

### reviews

In [17]:
# getting the reviews table
query = ''' 

SELECT *
FROM reviews
LIMIT 5 

'''
reviews = pd.io.sql.read_sql(query, con = engine) 
display(reviews)
reviews.dtypes

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

review_id     int64
book_id       int64
username     object
text         object
dtype: object

There are 4 columns in the table: `review_id` — the review ID; `book_id` — the book ID; `username` — the name of the user who wrote the review; `text` — the text of the review

## Analytical tasks

### How many books were published after January 1, 2000

In [18]:
query = ''' 

SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01' 

'''   
books_count = pd.io.sql.read_sql(query, con = engine) 
books_count

count
0    819

**819** books were published after January 1, 2000.

### The number of reviews and the average score for each book

In [19]:
query = ''' 

SELECT books.title,
       COUNT (distinct review_id) as review_count,
       AVG(rating) AS avg_rating
FROM books
LEFT JOIN reviews ON reviews.book_id = books.book_id
LEFT JOIN ratings ON ratings.book_id = books.book_id

GROUP BY books.book_id
ORDER BY COUNT(reviews.review_id) DESC

'''
reviews_and_avg_rating = pd.io.sql.read_sql(query, con = engine)
reviews_and_avg_rating

title  review_count  \
0                              Twilight (Twilight  #1)             7   
1                  The Hobbit  or There and Back Again             6   
2                               The Catcher in the Rye             6   
3    Harry Potter and the Prisoner of Azkaban (Harr...             6   
4    Harry Potter and the Chamber of Secrets (Harry...             6   
..                                                 ...           ...   
995                          Essential Tales and Poems             0   
996                               Leonardo's Notebooks             0   
997   The Cat in the Hat and Other Dr. Seuss Favorites             0   
998    Anne Rice's The Vampire Lestat: A Graphic Novel             0   
999                            The Natural Way to Draw             0   

     avg_rating  
0      3.662500  
1      4.125000  
2      3.825581  
3      4.414634  
4      4.287500  
..          ...  
995    4.000000  
996    4.000000  
997    5.000000  
998    3.666667  
999    3.000000  

[1000 rows x 3 columns]

We see that the maximum number of reviews is **7** for the book **Twiligh"**, then we see **Hobbit**, **The Catcher in the Rye** and the **Harry Potter** book with the 6 reviews. We also see the average ratings for each book.

### The publishing house that released the largest number of books with more than 50 pages

In [20]:
query = ''' 

SELECT publishers.publisher,

        COUNT(num_pages) as count_books
        
FROM books
INNER JOIN publishers on publishers.publisher_id = books.publisher_id
WHERE books.num_pages > 50
GROUP BY publishers.publisher
ORDER BY count_books DESC 
LIMIT 3

'''

publishers_big_books = pd.io.sql.read_sql(query, con = engine)
publishers_big_books

publisher  count_books
0             Penguin Books           42
1                   Vintage           31
2  Grand Central Publishing           25

The publishers that produce the most books with more than 50 pages are **Penguin Books**, **Vintage** and **Grand Central Publishing**.

### The author with the highest average rating of books (only books with 50 or more ratings)

In [21]:
query = ''' 

SELECT authors.author,
        AVG(ratings.rating) as mean_rating
FROM books
INNER JOIN authors on authors.author_id = books.author_id
INNER JOIN ratings on ratings.book_id = books.book_id
WHERE books.book_id IN (SELECT book_id
                        FROM ratings
                        GROUP BY book_id
                        HAVING COUNT(rating) >= 50)
GROUP BY authors.author
ORDER BY mean_rating DESC
LIMIT 1 


'''    
top_author = pd.io.sql.read_sql(query, con = engine)
top_author

author  mean_rating
0  J.K. Rowling/Mary GrandPré     4.287097

The author with the highest average book rating among books with 50 or more ratings is  **J.K. Rowling**, Mary GrandPré.

### The average number of reviews from users who have given more than 50 ratings

In [22]:
query = ''' 


SELECT SUM(group_table.count_rev)/COUNT(group_table.count_rev) as mean
        FROM (SELECT reviews.username, 
        COUNT(reviews.username) AS count_rev
FROM reviews
WHERE reviews.username IN (SELECT ratings.username
                            FROM ratings
                            GROUP BY ratings.username
                            HAVING COUNT(ratings.rating_id) > 50)
                  
                            GROUP BY username) AS group_table 
                            

'''         
review_avg_count = pd.io.sql.read_sql(query, con = engine)
review_avg_count

mean
0  24.333333

The average number of reviews from users who have given more than 50 ratings is **24**.

## Conclusion

Based on the results of the analysis, 5 tables from the database were derived and studied.

It was revealed:

1)  819 books were published after January 1, 2000.

2) The maximum number of reviews is 7 for the book "Twilight", Then we observe the Hobbit, "The Catcher in the Rye" and the Harry Potter book with 6 reviews. We also see average ratings for each book.

3) The publishers that produce the most books with more than 50 pages are Penguin Books, Vintage and Grand Central Publishing.

4) The author with the highest average book rating among books with 50 or more ratings is J.K. Rowling/Mary GrandPré.

5) The average number of reviews from users who have given more than 50 ratings is 24.

According to the results of the analysis, it is clear that users are willing to read fiction. At the same time, they actively leave reviews and evaluate books. Perhaps it would be a good idea to increase the variety of books in this genre, since in the pandemic people increasingly want to escape from reality. It would also be possible to work on related genres, such as detective fiction and so on.